packages/jupyter-ai/jupyter_ai/tools/models.py

```diff
@@ -0,0 +1,232 @@
+import re
+from typing import Callable, Optional
+
+from pydantic import BaseModel, ConfigDict, Field, model_validator
+
+
+def get_doc_description(func) -> str:
+    """Extract the first paragraph from a function's docstring.
+
+    Args:
+        func: The function to extract documentation from.
+
+    Returns:
+        The first paragraph of the function's docstring, cleaned up and
+        with whitespace normalized. Returns empty string if no docstring exists.
+
+    Example:
+        >>> def sample_func():
+        ...     '''This is the first paragraph.
+        ...
+        ...     This is the second paragraph.'''
+        ...     pass
+        >>> get_doc_description(sample_func)
+        'This is the first paragraph.'
+    """
+    if not func.__doc__:
+        return ""
+
+    # Split docstring into paragraphs
+    paragraphs = re.split(r"\n\s*\n", func.__doc__.strip())
+
+    # Return the first paragraph, cleaned up
+    if paragraphs:
+        return re.sub(r"\s+", " ", paragraphs[0].strip())
+
+    return ""
+
+
+class Tool(BaseModel):
+    """Represents a tool with callable function and capability metadata.
+
+    A Tool wraps a callable function with metadata including name, description,
+    and flags that indicate what operations the tool can perform.
+
+    Attributes:
+        callable: The function that implements the tool's functionality.
+        name: The tool's name. Auto-generated from callable if not provided.
+        description: Tool description. Auto-extracted from callable docstring if not provided.
+        read: Whether the tool can read data.
+        write: Whether the tool can write data.
+        execute: Whether the tool can execute operations.
+        delete: Whether the tool can delete data.
```

Comment:
[non-blocking] These docs can be made under each attribute; that also improves IDE intellisense since it will show the attribute's documentation when hovering over it.